internal/aws/xray/telemetry/sender.go

```diff
@@ -0,0 +1,288 @@
+	return pusherOptionFunc(func(r *telemetrySender) {
+		r.queueSize = queueSize
+	})
+}
+
+func WithBatchSize(batchSize int) Option {
+	return pusherOptionFunc(func(r *telemetrySender) {
+		r.batchSize = batchSize
+	})
+}
+
+type metadataProvider interface {
+	get() string
+}
+
+func getMetadata(providers ...metadataProvider) string {
+	var metadata string
+	for _, provider := range providers {
+		if metadata = provider.get(); metadata != "" {
+			break
+		}
+	}
+	return metadata
+}
+
+type simpleMetadataProvider struct {
+	metadata string
+}
+
+func (p simpleMetadataProvider) get() string {
+	return p.metadata
+}
+
+type envMetadataProvider struct {
+	envKey string
+}
+
+func (p envMetadataProvider) get() string {
+	return os.Getenv(p.envKey)
+}
+
+type ec2MetadataProvider struct {
+	client      *ec2metadata.EC2Metadata
+	metadataKey string
+}
+
+func (p ec2MetadataProvider) get() string {
+	var metadata string
+	if result, err := p.client.GetMetadata(p.metadataKey); err == nil {
+		metadata = result
+	}
+	return metadata
+}
+
+// ToOptions returns the metadata options if enabled by the config.
+func ToOptions(cfg Config, sess *session.Session, settings *awsutil.AWSSessionSettings) []Option {
+	if !cfg.IncludeMetadata {
+		return nil
+	}
+	metadataClient := ec2metadata.New(sess)
+	return []Option{
+		WithHostname(getMetadata(
+			simpleMetadataProvider{metadata: cfg.Hostname},
+			envMetadataProvider{envKey: envAWSHostname},
+			ec2MetadataProvider{client: metadataClient, metadataKey: metadataHostname},
+		)),
+		WithInstanceID(getMetadata(
+			simpleMetadataProvider{metadata: cfg.InstanceID},
+			envMetadataProvider{envKey: envAWSInstanceID},
+			ec2MetadataProvider{client: metadataClient, metadataKey: metadataInstanceID},
+		)),
+		WithResourceARN(getMetadata(
+			simpleMetadataProvider{metadata: cfg.ResourceARN},
+			simpleMetadataProvider{metadata: settings.ResourceARN},
+		)),
+	}
+}
+
+// NewSender creates a new Sender with a default interval and queue size.
+func NewSender(client awsxray.XRayClient, opts ...Option) Sender {
+	return newSender(client, opts...)
+}
+
+func newSender(client awsxray.XRayClient, opts ...Option) *telemetrySender {
+	sender := &telemetrySender{
+		client:    client,
+		interval:  defaultInterval,
+		queueSize: defaultQueueSize,
+		batchSize: defaultBatchSize,
+		stopCh:    make(chan struct{}),
+		Recorder:  NewRecorder(),
+	}
+	for _, opt := range opts {
+		opt.apply(sender)
+	}
+	return sender
+}
+
+// Start starts the loop to send the records.
+func (ts *telemetrySender) Start() {
+	ts.startOnce.Do(func() {
+		ts.stopWait.Add(1)
+		go func() {
+			defer ts.stopWait.Done()
+			ts.start()
+		}()
+	})
+}
+
+// Stop closes the stopCh channel to stop the loop.
+func (ts *telemetrySender) Stop() {
+	ts.stopOnce.Do(func() {
+		close(ts.stopCh)
+		ts.stopWait.Wait()
+	})
+}
+
+// start sends the queued records once a minute if telemetry data was updated.
+func (ts *telemetrySender) start() {
```

Comment:
```suggestion
// run sends the queued records once a minute if telemetry data was updated.
func (ts *telemetrySender) run() {
```

I think `run()` would be the name I'd expect most people to expect for a method like this.  It's unexported so not critical to change now but I had to read carefully when I saw `ts.Start()` call `ts.start()` and expect it to be long-running.